[AB3: Tracing Code] Add more headers to section instructions (#42)
se-edu/guides: tutorials/ab3TracingCode.md

# {{ title }}

<link rel="stylesheet" href="images/tracing/static/tracing.css">
> Indeed, the ratio of time spent reading versus writing is well over 10 to 1. We are constantly reading old code as part of the effort to write new code. …[Therefore,] making it easy to read makes it easier to write.
>
> —  Robert C. Martin Clean Code: A Handbook of Agile Software Craftsmanship

When trying to understand an unfamiliar code base, one common strategy used is to trace some representative execution path through the code base. One easy way to trace an execution path is to use a debugger to step through the code. In this tutorial, you will be using the IntelliJ IDEA’s debugger to trace the execution path of a specific user command.

> <h2 id="before-we-start">Before we start</h2>

Reviewer: No need. The anchor `before-we-start` is automatically generated by MarkBind. However, MarkBind will also warn about non-existent hash, which is a bug. Can ignore that for now.

Author: Noted

Now let’s set the breakpoint. First, double-click the item to reach the corresponding code. Once there, click on the left gutter to set a breakpoint, as shown below.
 ![LeftGutter](images/tracing/LeftGutter.png)

> 

Author: Removed

## Tracing the execution path

> At this point, you should have appreciated the general sequence diagram [**shown above**](#before-we-start)

Author: ```
At this point, you should have understood the high-level overview of the sequence diagram which shows the general execution flows between the main components [**shown above**](##setting-a-breakpoint).

We will begin to trace the following sequence diagram
which explores the `Logic` component in more detail which will be paramount to understanding other commands as well!
The diagrams will be reproduced with labels in their sections to highlight the pass of control between classes.

```

Made it clearer that its high level sequence diagram with all components and general execution flow and shifted it below so we can link to the low level sequence diagram with explores `Logic`

For this code tracing,
we will explore the `EditCommand` which is a feature that allows users to edit exising person fields in AB3.

<box type="tip" seamless>
    Recall from the <a href="https://se-education.org/addressbook-level3/UserGuide.html#editing-a-person--edit">User Guide</a> that the <code>edit</code> command has the format: <code>edit INDEX [n/NAME] [p/PHONE] [e/EMAIL] [a/ADDRESS] [t/TAG]…</code>.
</box>

For this tutorial, we will be issuing the command <code>edit 1 n/Alice Yeoh</code>

Specifically,
we trace the following sequence diagram
which explores the `Logic` component in more detail which will be paramount to understanding other commands as well!
The diagrams will be reproduced with labels in their sections to highlight the pass of control between classes.

> <puml src="images/tracing/LogicSequenceDiagramWithMainWindow.puml" alt="Tracing an `edit` command with specific arguments"></puml>

Author: Removed

<box type="tip" seamless>

**Tip:** Over the course of the debugging session, you will encounter every major component in the application. Try to keep track of what happens inside the component and where the execution transfers to another component.
</box>

### 1. MainWindow -> LogicManager

<annotate src="images/tracing/LogicSequenceDiagramWithMainWindow.png" width="900" alt="Sample Image">
  <!-- Set Legend to both -->

>   <a-point x="24%" y="16%"  label="1"/>

Reviewer: Should we use a different label e.g., `t1` to avoid confusing with step numbers in the explanations?

Author: Sure will call it T1

1. We are interested in the `logic.execute(commandText)` portion of that line so let’s _Step in_ into that method call:<br>
    ![StepInto](images/tracing/StepInto.png)

### 2. LogicManager -> AddressBookParser

<annotate src="images/tracing/LogicSequenceDiagramWithMainWindow.png" width="900" alt="Sample Image">

>   <!-- Set Legend to both -->

Reviewer: ```suggestion

```
Also update the `alt` property

  <a-point x="47.5%" y="23%"  label="2"/>
</annotate>

1. After stepping in, we note that `MainWindow` has passed control to the `Logic` component.

> 1. Specifically, We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.

Reviewer 2: Perhaps include `(not Logic#execute -- but this is expected because we know...)` as a sub-point for better readability? Then followed by `Given below is the same code...`

Author: Thanks!, will split into the two points

1. _Step over_ the logging code since it is of no interest to us now.
   ![StepOver](images/tracing/StepOver.png)

> 1. _Step into_ the line where user input in parsed from a String to a Command, which should bring you to the `AddressBookParser#parseCommand()` method (partial code given below):

Reviewer 2: Typo here.

```suggestion
1. _Step into_ the line where user input is parsed from a String to a Command, which should bring you to the `AddressBookParser#parseCommand()` method (partial code given below):
```

    ...
    ```

1. `EditCommandParser` is created first, annotated with [label 3](#3-addressbookparser-editcommandparser)

> 1. `AddressBookParser` now calls `parse(...)` of the newly created `EditCommandParser`, highlighted in the diagram as well, passing control to
>    `EditCommandParser`, annotated with [label 4](#3-addressbookparser-editcommandparser)

Reviewer 2: A suggestion for better readability, removing redundant whitespace.
```suggestion
1. `AddressBookParser` now calls `parse(...)` of the newly created `EditCommandParser`, which is highlighted in the diagram. This function call passes control to `EditCommandParser`, annotated with [label 4](#3-addressbookparser-editcommandparser)
```

### 4. EditCommandParser -> EditCommand

<annotate src="images/tracing/LogicSequenceDiagramWithMainWindow.png" width="900" alt="Sample Image">
  <!-- Set Legend to both -->
<a-point x="82%" y="51%"  label="5"/>
<a-point x="50%" y="64%"  color = "RED" label="6"/>

> <a-point x="26%" y="71%"  color = "RED" label="7"/>

Reviewer: Spacing around `=` is not consistent.

</annotate>

1. Let’s see what `EditCommandParser#parse()` does by stepping into it. You might have to click the 'step into' button multiple times here because there are two method calls in that statement: `EditCommandParser()` and `parse()`.

   <box type="tip" seamless>

   **Intellij Tip:** Sometimes, you might end up stepping into functions that are not of interest. Simply use the `step out` button to get out of them!
   </box>

1. Stepping through the method, `EditCommandParser#parse()`, shows that it calls `ArgumentTokenizer#tokenize()` and `ParserUtil#parseIndex()` to obtain the arguments and index required.

1. The rest of the method seems to exhaustively check for the existence of each possible parameter of the `edit` command and store any possible changes in an `EditPersonDescriptor`. Recall that we can verify the contents of `editPersonDesciptor` through the 'Variables' window.<br>
   ![EditCommand](images/tracing/EditCommandAnnotated.png)

> 1. Note the index and the contents of `EditPersonDescriptor`.
>    In this case, since the argument contains only name, only the `Name` field of
>    `EditPersonDescriptor` is set.

Reviewer 2: A suggestion to remove (redundant) whitespace.

```suggestion
1. Note the index and the contents of `EditPersonDescriptor`. In this case, since the argument contains only name, only the `Name` field of `EditPersonDescriptor` is set.
```

Author: Thanks

     <pic src="https://se-education.org/addressbook-level3/images/ModelClassDiagram.png" width="450" /><br>
   * {{ dg_ref }} This is a good time to read through the [**_Model component_** section of the DG](https://se-education.org/addressbook-level3/DeveloperGuide.html#model-component)

> 1. As you step through the rest of the statements in the `EditCommand#execute()` method,
>    you'll see that it creates a `CommandResult` object
>    (containing information about the result of the execution) and returns it to `LogicManager`.<br>
> 
> 

Reviewer 2: Removal of (redundant) whitespace.
```suggestion
1. As you step through the rest of the statements in the `EditCommand#execute()` method, you'll see that it creates a `CommandResult` object (containing information about the result of the execution) and returns it to `LogicManager`.<br>
```

   * {{ dg_ref }} This is a good time to read through the [**_Storage component_** section of the DG](https://se-education.org/addressbook-level3/DeveloperGuide.html#storage-component)

### 7. Returning to MainWindow

> 

Reviewer: ```suggestion

<!-- This puml tag forces MarkBind to generate the diagram (it's necessary as this diagram is not used in a puml tag anywhere else) -->
<puml style="display: none;" src="images/tracing/LogicSequenceDiagramOnlyWithMainWindow.puml"/>
```
